# Deep Learning with Python: Learning OR Operator

This notebook will help you understand what exactly is Deep Learning and how advancements
in industry has come when machines or computer programs are actually replacing humans. 
We will be covering:

1. Data Science and Its Components
2. The need of Deep Learning
3. What is Deep Learning?
4. Perceptrons and Artificial Neural Networks
5. Applications of Deep Learning?
6. Why Python for Deep Learning?
7. Deep Learning with Python: Perceptron Example
8. Deep Learning With Python: Creating a Deep Neural Network

## Data Science and It's Components

**Data Science** is the extaction of knowledge from data by using different techniques and algorithms.

**Artificial Intelligence** is a technique which enables machines to mimic human behavior. **Machine Learning** is
a subset of AI technique which uses statistical methods to enable machines to improve with experience. **Deep Learning**
is a subset of ML which make the computation of multi-layer neural network feasible. It uses Neural networks to simulate
human-like decision making.

## The need for Deep Learning

Machine Learning is based on the idea that machines should be given access to data and should
be left to learn and explore for themselves. It deals with the extraction of patterns from large
data sets. Handling large data sets was not a problem.
- Machine Learning Algorithms **cannot handle high-dimensional data**- where we have a large number
of inputs and outputs: around thousands of dimensions. Handling and processing such type of data 
becomes very complex and resource exhaustive. This is termed as **Curse of Dimensionality**.
- Another challenge faced was, to specify the **features to be extracted**. This plays an important
role in predicting the outcome as well as achieving better accuracy.

Deep Learning is **capable of handling the high dimensional data** and is also efficient in 
**focusing on the right features** on its own.

## What is Deep Learning?

Deep Learning is a subset of Machine Learning where similar Machine Learning Algorithms are used to train 
**Deep Neural Networks** so as to achieve better accuracy in those cases where the former was not performing up
to the mark. 

### Perceptrons and Artificial Neural Networks

An Artificial Neuron or a Perceptron is a linear model used for binary classification. The
neuron computes some function on these **weighted** inputs and gives the output.

It recieves n inputs then sums those inputs, applies a transformation and produces an output. It has two functions:
- Summation
- Transformation(Activation)

The weight shows the effectiveness of a particular input. 
**The more the weight of input, the more it will have an impact on the neural network**. On the other hand,
**Bias** is an additional parameter in the Perceptron which is used to adjust the output along with the weighted sum
of the inputs to the neuron which helps the model in a way that it can fit best for the given data.

**Activation Functions** translates the input into outputs. It uses a threshold to produce an output. There are many
functions that are used as Activation Functions, like:
- Linear or Identity
- Unit of Binary Step
- Sigmoid or Logistic
- Tanh
- ReLU
- Softmax

Note:
- Single-Layer Perceptrons **cannot classify non-linearly seperable data points**.
- Complex problems, that involve **a lot of parameters** cannot be solved by single-layer perceptrons.

A Neural Network is really just a **composition of Perceptrons, connected in different ways** and operating on
different activation functions.

## Applications of Deep Learning

Here are a few of the important ones that are present in our Day to Day tasks.

- Handwritten Digit Recognition 
- Machine Translation
- Facial Recognition and Automatic Tagging
- Virtual Personal Assistants
- Self Driving Car
- Chatbots

## Why Python for Deep Learning?

- Python is a **general purpose programming language** as being **easy to use** when it comes to analytical and 
quantitative computing.
- Python is **Dynamically Typed**
- Huge Community Support
- A vast range of Libraries for different purposes like **Numpy, Seaborn, Matplotlib, Pandas, and Scikit-learn**.

## Deep Learning with Python: Perceptron Example

We are going to work with the "OR" gate. The output is 1 if any of the inputs is also 1

| X1| X2| Y |
| - | - | - |
| 0 | 0 | 0 |
| 0 | 1 | 1 |
| 1 | 0 | 1 |
| 1 | 1 | 1 |

Therefore, a Perceptron can be used as a separator or a decision line that divides that input set of OR Gate, into
two classes. Inputs having output as 0 that lies below the decision line. Inputs having output as 1 that lies above the
decision line or separator.

Mathematically a perceptron can be though of like an equation of Weights $W$, Inputs $x$, and Bias $b$.
$$
f(x) = W\cdot x + b
$$

### Step 1: Import all the required library

In [1]:
import tensorflow as tf
tf.compat.v1.disable_eager_execution()

### Step 2: Define Vector Variables for Input and Output

In [2]:
train_in = [
[0,0,1],
[0,1,1],
[1,0,1],
[1,1,1]]
 
train_out = [
[0],
[1],
[1],
[1]]


### Step 3: Define Weight Variable

Here we will define the tensor variable of shape 3×1 for our weights and assign some random values to it initially.

In [3]:
w = tf.Variable(tf.random.normal([3, 1], seed=15))

### Step 4: Define placeholders for Input and Output

We need to define placeholders so that they can accept external inputs on the run.

In [4]:
x = tf.compat.v1.placeholder(tf.float32,[None,3])
y = tf.compat.v1.placeholder(tf.float32,[None,1])

### Step 5: Calculate Output and Activation Function

As discussed earlier, the input received by a perceptron is first multiplied by the respective weights and then, all these weighted inputs are summed together. This summed value is then fed to activation for obtaining the final result.

In [5]:
output = tf.nn.relu(tf.matmul(x, w))

### Step 6: Calculate the Cost or Error

In [6]:
loss = tf.reduce_sum(tf.square(output - y))

### Step 7: Minimize Error

The goal of a perceptron is to minimize the Loss or Cost or Error. So here we are going to use the Gradient Descent Optimizer.

In [7]:
optimizer = tf.compat.v1.train.GradientDescentOptimizer(0.01)
train = optimizer.minimize(loss)

### Step 8: Initialize all the variables

Variables are only defined with tf.Variable. So, we need to initialize the variables defined.

In [8]:
init = tf.compat.v1.global_variables_initializer()
sess = tf.compat.v1.Session()
sess.run(init)

2022-03-12 12:24:21.461945: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


### Step 9: Training Perceptron in Iterations

We need to train our perceptron i.e. update values of weights and bias in the successive iteration to minimize the error or loss. Here, I will train our perceptron in 100 epochs.

In [9]:
epochs = 100

for i in range(epochs):
    sess.run(train, {x:train_in,y:train_out})
    cost = sess.run(loss,feed_dict={x:train_in,y:train_out})
    print('Epoch--',i,'--loss--',cost)

Epoch-- 0 --loss-- 2.0738316
Epoch-- 1 --loss-- 1.7192812
Epoch-- 2 --loss-- 1.4468638
Epoch-- 3 --loss-- 1.2370586
Epoch-- 4 --loss-- 1.0750012
Epoch-- 5 --loss-- 0.94937146
Epoch-- 6 --loss-- 0.85154825
Epoch-- 7 --loss-- 0.7749659
Epoch-- 8 --loss-- 0.7146225
Epoch-- 9 --loss-- 0.66670835
Epoch-- 10 --loss-- 0.6283201
Epoch-- 11 --loss-- 0.59724545
Epoch-- 12 --loss-- 0.5717979
Epoch-- 13 --loss-- 0.5506916
Epoch-- 14 --loss-- 0.5329456
Epoch-- 15 --loss-- 0.5178112
Epoch-- 16 --loss-- 0.5047164
Epoch-- 17 --loss-- 0.49322423
Epoch-- 18 --loss-- 0.48300037
Epoch-- 19 --loss-- 0.47378856
Epoch-- 20 --loss-- 0.4653922
Epoch-- 21 --loss-- 0.4576601
Epoch-- 22 --loss-- 0.4504757
Epoch-- 23 --loss-- 0.44374895
Epoch-- 24 --loss-- 0.43740994
Epoch-- 25 --loss-- 0.43140402
Epoch-- 26 --loss-- 0.42568845
Epoch-- 27 --loss-- 0.42022946
Epoch-- 28 --loss-- 0.41500017
Epoch-- 29 --loss-- 0.40997905
Epoch-- 30 --loss-- 0.40514854
Epoch-- 31 --loss-- 0.4004942
Epoch-- 32 --loss-- 0.39600414
Epoc

### As you can see here, the loss started at **2.07** and ended at **0.27**!